### DISCUSSION TITLE
How to size an image to best fit inside its parent

### DISCUSSION THREAD

**Discussion by ElliotEserinUT** [AvaloniaUI/Avalonia#15526] [Q&A]
I have been trying for a while to figure out how to properly scale an image control so it fits inside its parent. This is maybe best explained with an example:
I have a DockPanel, and then I have an image which is not docked so it should fit the remaining space. But the image always fits into the container based on its width and not height, which means the bottom of the image will be cropped in certain aspect ratios, or there is excess whitespace in others. If I scale the window horizontally the image shrinks, but if I scale the window vertically, the image will not resize to fill the smaller space. How might I fix this?
Currently this is all my code does:
```
	<DockPanel>
		<StackPanel DockPanel.Dock="Top">
			...
		</StackPanel>
		<Image Source="{Binding Image}" Stretch="Uniform" ClipToBounds="True" />
	</DockPanel>
```

But I have tried a few variations using panels and verticalAlignment and whatnot and nothing seems to work...
Thanks in advance!

**Comment by packdat** [comment]
I'm writing this for anyone else like me who just started to experiment with Avalonia UI.
Building my very first Avalonia app i got stuck with the exact same problem and searching the web lead me to this issue.

Try to put your image into something different than a `StackPanel`. 

I had a the following initial setup, pretty standard stuff, a grid with 3 rows (Toolbar, Work Area, Statusbar):
```
    <Grid>
        <Grid.RowDefinitions>
            <RowDefinition Height="Auto"></RowDefinition>
            <RowDefinition Height="*"></RowDefinition>
            <RowDefinition Height="Auto"></RowDefinition>
        </Grid.RowDefinitions>
        <!-- Toolbar -->
        <StackPanel Orientation="Horizontal" Margin="8" Spacing="10" Grid.Row="0">
            ...
        </StackPanel>
        <StackPanel Grid.Row="1">
            <Image x:Name="VideoFrame" Stretch="Uniform" ClipToBounds="True" 
                   VerticalAlignment="Center" HorizontalAlignment="Center"></Image>
        </StackPanel>
        <!-- Status bar -->
        <StackPanel Orientation="Horizontal" Margin="8" Spacing="10" Grid.Row="2">
            ...
        </StackPanel>
    </Grid>
```
The image itself was generated in code-behind using a `WriteableBitmap`.
Upon resizing the window, the image was "leaking" into the status-bar below it and i couldn't figure out how to prevent that with the properties available for `StackPanel`.

Then i decided to try something simpler than `StackPanel`, as the `Image` was the only control inside it.
I switched from `StackPanel` to `Border` and things started to work as expected:
```
    <Grid>
        <Grid.RowDefinitions>
            <RowDefinition Height="Auto"></RowDefinition>
            <RowDefinition Height="*"></RowDefinition>
            <RowDefinition Height="Auto"></RowDefinition>
        </Grid.RowDefinitions>
        <!-- Toolbar -->
        <StackPanel Orientation="Horizontal" Margin="8" Spacing="10" Grid.Row="0">
            ...
        </StackPanel>
        <!-- Main area (Note: StackPanel wont scale/position the image properly !) -->
        <Border Grid.Row="1">
            <Image x:Name="VideoFrame" Stretch="Uniform" ClipToBounds="True" Margin="8"
                   VerticalAlignment="Center" HorizontalAlignment="Center"></Image>
        </Border>
        <!-- Status bar -->
        <StackPanel Orientation="Horizontal" Margin="8" Spacing="10" Grid.Row="2">
            ...
        </StackPanel>
    </Grid>
```
The image was now respecting it's parent's bounds and stayed nicely centered when resizing the window.
As mentioned, i just started to play around with Avalonia, there may be better and simpler ways to approach this.

**Comment by RobG66** [comment]
Why not wrap the image control in a viewbox?  

Rob

**Comment by Binary-Blaze** [comment]
I also use a viewbox for this.

`<Grid>

        <Grid.RowDefinitions>
            <RowDefinition Height="Auto"></RowDefinition>
            <RowDefinition Height="*"></RowDefinition>
            <RowDefinition Height="Auto"></RowDefinition>
        </Grid.RowDefinitions>
        
        <StackPanel Orientation="Horizontal" Margin="8" Spacing="10" Grid.Row="0">
            ...
        </StackPanel>
        
        <Viewbox Grid.Row="1">
            <Image Source="{Binding Image}"  />
        </Viewbox>
        
        <StackPanel Orientation="Horizontal" Margin="8" Spacing="10" Grid.Row="2">
            ...
        </StackPanel>

 </Grid>
`
